Avoid negative integer overflow when `filesize < io_->tell()`.

kevinbackhouse opened the pull request:
This fixes #791 on the master branch.

It is mostly the same as the fix for v0.27: #797.

This fix is nicer because I have added a new method named `readOrThrow` to `BasicIo`. But I couldn't make the same change in v0.27 because it changes the API of `BasicIo`.

kevinbackhouse committed:
Avoid negative integer overflow when `filesize < io_->tell()`.

This fixes #791.

piponazo commented:
Hi @kevinbackhouse. Just in case you did not notice, the windows build is broken after these changes:

```
..\src\webpimage.cpp(515): error C2220: warning treated as error - no 'object' file generated
..\src\webpimage.cpp(515): warning C4018: '<=': signed/unsigned mismatch
```

Do you have a windows development environment where to check/fix this? Otherwise I could take a look.

kevinbackhouse committed:
Fix signed/unsigned mismatch.

D4N commented on src/webpimage.cpp:
```diff
@@ -508,10 +509,11 @@ namespace Exiv2 {
             io_->readOrThrow(size_buff, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
-            const uint32_t size = Exiv2::getULong(size_buff, littleEndian);
+            const long size = Exiv2::getULong(size_buff, littleEndian);
+            enforce(0 <= size, Exiv2::kerCorruptedMetadata);
```
This check relies on undefined behavior (signed integer overflow) and is thus not effective, as the compiler is free to optimize it away.

D4N commented on src/webpimage.cpp:
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
+        enforce(0 <= filesize, Exiv2::kerCorruptedMetadata);
+        enforce((size_t)filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
```
```suggestion
        enforce(static_cast<size_t>(filesize) <= io_->size(), Exiv2::kerCorruptedMetadata);
```

D4N commented on src/webpimage.cpp:
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
 
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
```
Add a `#include "safe_op.hpp"` at the top of the file and remove the check in the following line. `Safe:add` catches overflow errors for you:
```suggestion
        const long filesize = Safe::add(Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian), 8);
```

D4N commented on src/webpimage.cpp:
```diff
@@ -489,15 +489,16 @@ namespace Exiv2 {
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
+        enforce(0 <= filesize, Exiv2::kerCorruptedMetadata);
```
```suggestion
```

D4N reviewed (dismissed):
I really like your proposed new API, but I am not so sure whether I really like the current interface.

D4N commented:
@kevinbackhouse Thanks for stepping forward and proposing this new API!

Since this is quite an important change, I'd like to discuss the API design first:
- we should be using `size_t` instead of `long` for the size parameter (depends on #786)
- the proposed API uses `DataBuf`, which we try to get rid off, so I'd prefer not to introduce a new API function that returns it
- I don't think the exception code should be a parameter that the API consumer can set, imho a generic read error would be fine

I've been thinking about something comparable and the best I came up with is the following:
``` C++
/// Read nbytes bytes into `dest` (enlarging it as needed)
/// @return a slice to the read chunk
[[nodiscard]] Slice<std::vector<uint8_t>> read(std::vector<uint8_t>& dest, size_t nbytes);
```

Provided you use the returned slice to access the data, this wouldn't even have to throw an exception on reaching `EOF`, as the slice will automatically throw if you perform an out of bound access. However, it might be useful to have this function "autothrow" on EOF (maybe make this configurable)?

Thoughts/opinions?

D4N committed:
Update src/webpimage.cpp

Co-Authored-By: kevinbackhouse <kev@semmle.com>

D4N committed:
Update src/webpimage.cpp

Co-Authored-By: kevinbackhouse <kev@semmle.com>

kevinbackhouse committed:
Remove unused Databuf API.

kevinbackhouse committed:
Improve portability of bounds checks.

kevinbackhouse committed:
Small change to comment.

kevinbackhouse committed:
Remove ErrorCode parameter from readOrThrow.

kevinbackhouse commented:
@D4N: I think all the signed/unsigned casts are fixed now. (Also in #797.)

I removed the `DataBuf` API since I wasn't actually using it. I only added it for completeness, so there's no need if it's going to be deprecated soon.

I removed the `ErrorCode` parameter from the API. I replaced it with always throwing `kerInputDataReadFailed`, because that sounded the most appropriate to me for a generic error message.

Wouldn't the simplest replacement for the `DataBuf` API be something like this?

```
std::vector<uint8_t> readOrThrow(size_t nbytes);
```

Note: I don't know the history of the `DataBuf` API or the reasons why you are trying to get rid of it.

kevinbackhouse commented:
I see that some of the builds are failing now due to clang's `-Wtautological-constant-out-of-range-compare` warning complaining about this line:

```
enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()), ...
```

How do you want me to fix that?

D4N commented:
> @D4N: I think all the signed/unsigned casts are fixed now. (Also in #797.)

Indeed, LGTM now.

> 
> I removed the `DataBuf` API since I wasn't actually using it. I only added it for completeness, so there's no need if it's going to be deprecated soon.
> 
> I removed the `ErrorCode` parameter from the API. I replaced it with always throwing `kerInputDataReadFailed`, because that sounded the most appropriate to me for a generic error message.

Sounds good to me.

> 
> Wouldn't the simplest replacement for the `DataBuf` API be something like this?
> 
> ```
> std::vector<uint8_t> readOrThrow(size_t nbytes);
> ```

Yes, but this has the disadvantage that you always get a heap allocation and cannot start reading at an arbitrary position inside a user provided buffer (although that also doesn't work with my proposed API).

> 
> Note: I don't know the history of the `DataBuf` API or the reasons why you are trying to get rid of it.

`DataBuf` is just an inferior version of `std::vector`: it provides raw access to the underlying array, it has no automatic bounds checking (like `std::vector::at()`), it's memory management is relatively questionable, uses `long` where one should use `size_t`, etc.pp.

> I see that some of the builds are failing now due to clang's `-Wtautological-constant-out-of-range-compare` warning complaining about this line:
> 
> ```
> enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()), ...
> ```
> 
> How do you want me to fix that?

Do you want a quick fix or a proper one?

A quick fix is to include that code only on Windows, as every other platform has `sizeof(long) > sizeof(uint32_t)` and there the above code is a NOP.

A proper fix would be that I or `$volunteer` implements `Safe::cast<T>`, which has been on my todo list for quite some time.

D4N commented on src/webpimage.cpp:
```diff
@@ -522,7 +522,7 @@ namespace Exiv2 {
 
             // Check that `size_u32` is safe to cast to `long`.
-            enforce(size_u32 <= static_cast<size_t>(std::numeric_limits<long>::max()),
+            enforce(static_cast<size_t>(size_u32) <= static_cast<size_t>(std::numeric_limits<long>::max()),
```
Wrap the enforce in a 
```C++
#if defined(_WIN32) || defined(__CYGWIN__) || defined(__MINGW__) || defined(_MSC_VER)
enforce(...);
#endif
```

kevinbackhouse replied:
Thanks, I'll do that if I cannot defeat the compiler warning. I have one more thing that I will try first.

kevinbackhouse committed:
Use the preprocessor to fix the compiler warning.

kevinbackhouse committed:
Use the preprocessor to fix the compiler warning.
